device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/SasTokenProvider.java

```diff
@@ -0,0 +1,34 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.microsoft.azure.sdk.iot.device;
+
+/**
+ * Interface for allowing users to control SAS token generation. To see an example of how SAS tokens can be generated
+ * from device connection strings, see {@link com.microsoft.azure.sdk.iot.device.auth.IotHubSasToken}.
+ * @see <a href="https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#example">This document</a> for more details on sas tokens.
+ */
+public interface SasTokenProvider
+{
+    /**
+     * Returns a SAS token to be used for authentication.
+     * <p>
+     * This function will be called each time the client library needs a SAS token. This will follow different patterns
+     * based on which protocol your client object is using.
+     * <p>
+     * For HTTPS, this function will be called for each HTTPS request made (for instance, once per telemetry send), and does not need
+     * to be a unique token each time. Because of that, users may simply cache and return the same SAS token repeatedly here.
+     * Alternatively, users can generate very short-lived SAS tokens since they will only need to be valid for a relative short period of time.
+     * The user of this API is responsible for tracking when to renew the SAS token based on how long the previous token was valid for.
+     * <p>
+     * For AMQPS/AMQPS_WS, this function will be called once when first opening the connection, and then will be called again
+     * at some point prior to the previous SAS token's expiry time in order to proactively renew the connection's authentication.
+     * This proactive renewal takes place at around 85% of the previous SAS token's lifespan.
```

Comment:
If I understand coreectly, for a user implementing this interface, they would need to implement their own renewal logic, right? This statement is only for our library's implementation of `SasTokenProvider`, which incidentally contains the 85% lifetime renewal logic?

Correct me if I am wrong, this is what I understand - 
From an interface perspective, our library calls `getSasToken()` whenever it requires a new SAS token (at it's simplest implementation - every time for Http, on CONNECT for Mqtt, while opening cbs link for Amqp).
For a user of this interface, you can then add any caching or renewal logic as you require in your implementation of `getSasToken()`, while the sdk can add additional calls to `getSasToken()` on reconnection, with an amqp refresh timer, etc.

If this is correct, then we should instead state that any renewal or caching logic is the ownership of whoever implements `SasTokenProvider` interface.

Reply:
Wait, I see a lot of authentication provider classes in our library; would you mind listing their hierarchy? That will help understand which method (in the auth provider interface) our transport layer calls into.

Reply:
If `SasTokenProvider` is not what our transport layer calls into, then my assumption above might not be correct.

Reply:
I'd need a whiteboard to illustrate these authentication providers and their hierarchies. Just focusing on the SAS side of the tree, IotHubSasTokenAuthenticationProvider is the "root" class for all symmetric key based authentication mechanisms. From there, there are several implementations depending on if the device key is present, or if the TPM is the provider, or if the new SAS token provider interface is provided.

Reply:
I once attempted to draw the hierarchy by just looking at the code and I failed on the first few attempts ... it's a confusing logic.

Reply:
> This statement is only for our library's implementation of SasTokenProvider, which incidentally contains the 85% lifetime renewal logic?

Our SDK will still ask for a new SAS token once the previous token has 15% of it's life left, but the SDK doesn't mandate anything about that token. It's not recommended, but customers could just return the same SAS token at that point, for instance.